src/coreclr/jit/optimizer.cpp

```diff
@@ -7608,12 +7608,791 @@ bool Compiler::optIsRangeCheckRemovable(GenTree* tree)
+
+            foldOp = GT_OR;
+            cmpOp  = GT_NE;
+        }
+    }
+    else
+    {
+        // The b1 condition must be the reverse of the b2 condition because the only operators
+        // that we will see here are GT_EQ and GT_NE. So, if they are not the same, we have one of each.
+
+        if (pOptBoolsDsc->t1.compTree->gtOper == pOptBoolsDsc->t2.compTree->gtOper)
+        {
+            return changeCond;
+        }
+
+        if (pOptBoolsDsc->t1.compTree->gtOper == GT_EQ)
+        {
+            // t1:c1==0 t2:c2!=0 ==> Branch to BX if both values are non-0
+            // So we will branch to BX if (c1&c2)!=0
+
+            foldOp = GT_AND;
+            cmpOp  = GT_NE;
+        }
+        else
+        {
+            // t1:c1!=0 t2:c2==0 ==> Branch to BX if both values are 0
+            // So we will branch to BX if (c1|c2)==0
+
+            foldOp = GT_OR;
+            cmpOp  = GT_EQ;
+        }
+    }
+
+    // Anding requires both values to be 0 or 1
+
+    if ((foldOp == GT_AND) && (!pOptBoolsDsc->t1.isBool || !pOptBoolsDsc->t2.isBool))
+    {
+        return changeCond;
+    }
+
+    //
+    // Now update the trees
+    //
+
+    pOptBoolsDsc->foldOp   = foldOp;
+    pOptBoolsDsc->foldType = foldType;
+    pOptBoolsDsc->cmpOp    = cmpOp;
+
+    optOptimizeBoolsUpdateTrees(pOptBoolsDsc);
+
+    // Update the change to true to continue the bool optimization for the rest of the BB chain
+    changeCond = true;
+
+#ifdef DEBUG
+    if (verbose)
+    {
+        printf("Folded %sboolean conditions of " FMT_BB " and " FMT_BB " to :\n", c2->OperIsLeaf() ? "" : "non-leaf ",
+               pOptBoolsDsc->b1->bbNum, pOptBoolsDsc->b2->bbNum);
+        gtDispStmt(s1);
+        printf("\n");
+    }
+#endif
+
+    return changeCond;
+}
+
+//
+// optOptimizeBoolsChkBlkCond: Checks block conditions if it can be boolean optimized
+//
+// Arguments:
+//      pOptBoolsDsc    The descriptor with Basic Block b1, b2 and b3 set when the method is called
+//
+// Return:
+//      s1              If all conditions pass, returns the last statement of b1.
+//
+// Notes:
+//      This method checks if the second or third block contains only one statement, and
+//      checks if tree operators are of the right type, e.g, GT_JTRUE, GT_RETURN.
+//
+//      When called, b1, b2 and b3 of pOptBoolsDsc are set, where
+//          b1 is the first block, b2 is the second block and b3 is the third block.
+//      If it passes all the conditions, t1, t2 and t3 are set to the root nodes of b1, b2 and b3 each.
+//      SameTarget is also updated to true if b1 and b2 jump to the same destination.
+//
+Statement* Compiler::optOptimizeBoolsChkBlkCond(OptBoolsDsc* pOptBoolsDsc)
+{
+    BasicBlock* b1 = pOptBoolsDsc->b1;
+    BasicBlock* b2 = pOptBoolsDsc->b2;
+    BasicBlock* b3 = pOptBoolsDsc->b3;
+
+    bool optReturnBlock = false;
+    if (b3)
+    {
+        optReturnBlock = true;
+    }
+
+    // Check jump destination condition
+
+    if (!optReturnBlock)
```

Comment:
It seems like this code should be hoisted all the way out to `optOptimizeBools`: it is specific to the two cases handled, and there is no shared functionality. It would save some function calls before exiting.

Reply:
> there is no shared functionality

@BruceForstall I don't understand what you mean by this. I refactored the code and moved this code into `optOptimizeBoolsChkBlkCond()` in case boolean optimization will be extended for other patterns.

Reply:
Done. Moved the code that checks the jump destination to `optOptimizeBools`.